core/directory.go

```diff
@@ -303,43 +298,38 @@ func (dir *Directory) Glob(ctx context.Context, src string, pattern string) ([]s
 	}
 	// empty directory, i.e. llb.Scratch()
 	if ref == nil {
-		if clean := path.Clean(dir.Dir); clean == "." || clean == "/" {
-			return []string{}, nil
-		}
-		return nil, fmt.Errorf("%s: no such file or directory", src)
+		return []string{}, nil
 	}
 
-	entries, err := ref.ReadDir(ctx, bkgw.ReadDirRequest{
-		Path:           path.Join(dir.Dir, src),
-		IncludePattern: pattern,
-	})
+	// We use the same pattern matching function as Buildkit, since just doing
+	// IncludePatterns will still include directories we don't want
+	pm, err := patternmatcher.New([]string{pattern})
 	if err != nil {
 		return nil, err
 	}
 
-	paths := []string{}
-	for _, entry := range entries {
-		entryPath := path.Join(src, entry.GetPath())
-
-		// We use the same pattern matching function as Buildkit to handle
-		// recursive strategy.
-		match, err := patternmatcher.MatchesOrParentMatches(entryPath, []string{pattern})
-		if err != nil {
-			return nil, err
-		}
-
-		if match {
-			paths = append(paths, entryPath)
-		}
-
-		// Handle recursive option
-		if entry.IsDir() {
-			subEntries, err := dir.Glob(ctx, entryPath, pattern)
+	var paths []string
+	err = ref.WalkDir(ctx, buildkit.WalkDirRequest{
+		IncludePattern: pattern,
+		Path:           dir.Dir,
+		Callback: func(path string, info *fstypes.Stat) error {
+			// HACK: ideally, we'd have something like MatchesExact, which
+			// would skip the parent behavior that we don't really want here -
+			// oh well, let's just fake it with false
+			//nolint:staticcheck
+			match, err := pm.MatchesUsingParentResult(filepath.Clean(path), false)
```

Comment:
We might be able to replace this with https://github.com/moby/patternmatcher/pull/3, if that gets merged :tada: